### ISSUE TITLE
Usage of `BitField::load` is unintuitive when slicing across bit store boundary on little-endian machines

### ISSUE THREAD

**Issue by hcsch** [bitvecto-rs/bitvec#150]
During my usage of this crate for [day 16 of the Advent of Code](https://adventofcode.com/2021/day/16), I stumbled across some unintuitive / incorrect behavior exhibited by the `BitField::load` function, where the third of the following asserts fails:
```rust
assert_eq!(bits![u8, Msb0; 0, 0, 1].load::<u8>(), 1u8); // ok
assert_eq!(
    bits![u8, Msb0; 0, 0, 0, 0, 0, 0, 0, 0, 1][6..],
    bits![u8, Msb0; 0, 0, 1]
); // ok
assert_eq!(
    bits![u8, Msb0; 0, 0, 0, 0, 0, 0, 0, 0, 1][6..].load::<u8>(),
    1u8
); // assertion failed: `(left == right)` left: `4`, right: `1`
```
As suggested by the value I filled in for the right value of the assert, I expect this `load` to return `1`. The actual behavior however is that it returns `4`.

Especially since `bits![u8, Msb0; 0, 0, 0, 0, 0, 0, 0, 0, 1][6..],` is equal to `bits![u8, Msb0; 0, 0, 1]`, I expect the `load` function to return the same value for both of them.

If I have misunderstood the usage of `BitField::load`, I would appreciate if you could help me understand how `load` actually works.

*NOTE: in my actual code I use a `BitVec<usize, Msb0>`, and this issue occurs slicing with `[62..]` and then `load`ing three bits into a `u8`.*

I encountered this issue with version `1.0.0-rc1` of this crate.

**Comment by hcsch** [comment]
I'm sorry for the initially rather dramatic title. After looking into this a bit further, it seems that the issue was with endianness.
Intuitively, I'd expect loading a value out of a `BitSlice<I, Msb0>` to work as the `BitOrder` suggests, regardless of the endianness of the machine the code runs on. I might have paid more attention to endianness had I created a `BitVec<I, Msb0>` from an iterator of bit store integers, as opposed to from an iterator of `bool`s, which is how I'm doing it currently in the hopes of being able to avoid having to think about endianness and bit store boundaries.

I assume I'll just have to think about endianness and storage boundaries after all and use `u8`s, big-endianness, and the corresponding `load_be` function. I'll close this issue for now, but if anybody has some insights to share, feel free to do so.

**Comment by myrrlyn** [comment]
This is a *really* common pitfall and unfortunately I just haven't found a way to make it work correctly in every possible case. It's important for wire-format use cases that `BitField` have as full as possible a matrix of storage representations, but unfortunately this does, as you have noted, lead to discrepancies between bit-sequence literals and the numeric values they purport to represent.

In the v1 release candidate series I have attempted to write [the documentation](https://docs.rs/bitvec/1.0.0-rc1/bitvec/field/index.html) much more thoroughly to indicate that `BitField` can have aberrant behavior, in particular, [here](https://docs.rs/bitvec/1.0.0-rc1/bitvec/field/index.html#bit-slice-storage-and-integer-value-relationships):

> If you do not want to introduce the processor’s byte endianness as a variable that affects the in-memory representation of stored integers, use `BitSlice<u8, _>` as the bit-field storage type. In particular, `BitSlice<u8, Msb0>` will fill memory in a way that intuitively matches what most debuggers show when inspecting memory.

and in the [## Endianness](https://docs.rs/bitvec/1.0.0-rc1/bitvec/field/index.html#endianness) section. The specific documentation for [`BitField::load_be`](https://docs.rs/bitvec/1.0.0-rc1/bitvec/field/trait.BitField.html#tymethod.load_be) also describes the behavior used when translating memory elements into numerical values.

I have tried to figure out what a reasonable default behavior here is and I've settled on prioritizing round-tripping a numerical value from store to load over tracking what a bit-sequence literal might look like.

Unfortunately the only real help I can give here is, as you've already noted, "use `<u8, Msb0>` and spend more time in the docs". It's an unfortunately complex problem and the tension between default-behavior requirements for the user stories of "I want to store individual bits" vs "I want to store numbers in unusual ways" is not really reconcilable.

Thank you for reporting; I really do appreciate feedback on usability even if I'm not around to look at it right away and you wind up rubber-ducking here.
